File: src/main/java/recipeapplication/utility/IngredientConverter.java
package recipeapplication.utility;

public class IngredientConverter
{
    private static final double OUNCE_TO_GRAM = 28.35;
    private static final double KILOGRAM_TO_POUND = 2.205;
    private static final String OUNCE = "oz";
    private static final String GRAM = "g";
    private static final String KILOGRAM = "kg";
    private static final String POUND = "lbs";

    public static String toImperial(String quantity)
    {
        if (quantity == null)
        {
            return null;
        }

        String[] words = quantity.split(" ");

        int index = 0;
        for (String word : words)
        {
            try
            {
                if (word.toLowerCase().endsWith(KILOGRAM))
                {
                    double value = extractValue(word, KILOGRAM);
                    value = value * KILOGRAM_TO_POUND;
                    words[index] = roundOneDecimalPlace(value) + POUND;

                } else if (word.toLowerCase().endsWith(GRAM))
                {
                    double value = extractValue(word, GRAM);
                    value = value / OUNCE_TO_GRAM;
                    words[index] = roundOneDecimalPlace(value) + OUNCE;
                }
            }
            catch (NumberFormatException e)
            {
                // If a number cannot be extracted we move on to the next word
            }

            index++;
        }

        return String.join(" ", words);
    }

    public static String toMetric(String quantity)
    {
        if (quantity == null)
        {
            return null;
        }

        String[] words = quantity.split(" ");

        int index = 0;
        for (String word : words)
        {
            try
            {
                if (word.toLowerCase().endsWith(POUND))
                {
                    double value = extractValue(word, POUND);
                    value = value / KILOGRAM_TO_POUND;
                    words[index] = roundOneDecimalPlace(value) + KILOGRAM;

                } else if (word.toLowerCase().endsWith(OUNCE))
                {
                    double value = extractValue(word, OUNCE);
                    value = value * OUNCE_TO_GRAM;
                    value = Math.round(value);
                    words[index] = ((int) value) + GRAM;
                }
            }
            catch (NumberFormatException e)
            {
                // If a number cannot be extracted we move on to the next word
            }

            index++;
        }

        return String.join(" ", words);
    }

    private static double extractValue(String word, String measurement)
    {
        return Double.parseDouble(word.substring(0, word.length() - measurement.length()));
    }

    private static double roundOneDecimalPlace(double value)
    {
        return Math.round(value * 10.0) / 10.0;
    }
}

File: src/main/java/recipeapplication/utility/IngredientConverter.kt
package recipeapplication.utility

import kotlin.math.roundToInt

private const val OUNCE_TO_GRAM = 28.35
private const val KILOGRAM_TO_POUND = 2.205
private const val OUNCE = "oz"
private const val GRAM = "g"
private const val KILOGRAM = "kg"
private const val POUND = "lbs"

fun toImperial(ingredient: String) : String {
    return convertMeasurements(ingredient) { word ->
        when (true) {
            word.toLowerCase().endsWith(KILOGRAM) -> {
                val value = (word extractValue  KILOGRAM) * KILOGRAM_TO_POUND
                roundOneDecimalPlace(value).toString() + POUND
            }
            word.toLowerCase().endsWith(GRAM) -> {
                val value = (word extractValue  GRAM) / OUNCE_TO_GRAM
                roundOneDecimalPlace(value).toString() + OUNCE
            }
            else -> word
        }
    }
}

fun toMetric(ingredient: String): String {
    return convertMeasurements(ingredient) { word ->
        when (true) {
            word.toLowerCase().endsWith(POUND) -> {
                val value = (word extractValue  POUND) / KILOGRAM_TO_POUND
                roundOneDecimalPlace(value).toString() + KILOGRAM
            }
            word.toLowerCase().endsWith(OUNCE) -> {
                val value = ((word extractValue OUNCE) * OUNCE_TO_GRAM).roundToInt()
                value.toString() + GRAM
            }
            else -> word
        }
    }
}

private infix fun String.extractValue(measurement: String): Double {
    return this.substring(0, this.length - measurement.length).toDouble()
}

private fun roundOneDecimalPlace(value: Double): Double {
    return (value * 10.0).roundToInt() / 10.0
}

private fun convertMeasurements(ingredient: String, convert: (String) -> String) : String {
    val words = ingredient.split(" ")
    return words.joinToString(separator = " ") { word ->
        try {
            convert(word)
        } catch (e: NumberFormatException) {
            word
        }
    }
}
